ci/clean_notebooks.py

```diff
@@ -38,27 +42,28 @@ def clean_notebook(file: pathlib.Path, check_only=False) -> None:
     # Remove the output and metadata from each cell
     # also reset the execution count
     # if the cell has no code, remove it
+    fields_defaults: List[Tuple[str, Any]] = [
+        ("execution_count", None),
+        ("outputs", []),
+        ("metadata", {}),
+    ]
     for cell in nb.cells:
-        if "outputs" in cell and cell["outputs"]:
-            if check_only:
-                raise UncleanNotebookError(f"Notebook {file} has outputs")
-            cell["outputs"] = []
-            was_dirty = True
-        if "metadata" in cell and cell["metadata"]:
-            if check_only:
-                raise UncleanNotebookError(f"Notebook {file} has metadata")
-            cell["metadata"] = {}
-            was_dirty = True
-        if "execution_count" in cell and cell["execution_count"]:
-            if check_only:
-                raise UncleanNotebookError(f"Notebook {file} has execution count")
-            cell["execution_count"] = None
-            was_dirty = True
         if cell["cell_type"] == "code" and not cell["source"]:
             if check_only:
                 raise UncleanNotebookError(f"Notebook {file} has empty code cell")
             nb.cells.remove(cell)
             was_dirty = True
+        for field, default in fields_defaults:
+            if cell.get(field) != default:
```

Comment:
There's a slight semantic change here which I think wasn't intended. Previously you were OK with `"field" not in cell"` -- but now for outputs/metadata you add the default. This has caused the notebooks to have `"outputs": []` even for text entries that don't usually produce any output.

Probably want to change this to something like `if field in cell and cell.get(field) != default`? Or have a list of permissible values as well as defaults, like:

```python
for field, default, allowed in fields_defaults:
    if cell.get(field) not in allowed:
        # ....
```

where you could include `None` in `allowed` for `"outputs"` even while keeping `[]` as the default.

Reply:
I think the "correct" way to format is to have an empty list for code cells, no entry for any other cell. Not sure if the best way to do this is by allowing different defaults by different cell types. What you're suggesting achieves the same result when using a notebook executor because md cells never add the "outputs" key, but it would not roll it back automatically from e.g. the state it is now.

Reply:
Came up with a solution for this, it is somewhat different to what we've discussed, let me know if you want to take a look before I merge.